File: apps/android/app/src/main/java/com/exponential/app/domain/MergeTarget.kt
package com.exponential.app.domain

import com.exponential.app.data.db.CodingSessionEntity
import com.exponential.app.data.db.IssueEntity

/**
 * What a run's Merge control actually merges (EXP-734).
 *
 * Until now every merge went through an ISSUE — the run's own issue, or, for
 * a batch run, the representative issue of its combined PR (EXP-535). An
 * action or chat run has neither, and since EXP-626 it can still open a PR of
 * its own (`exponential_pr_open({repositoryId, head})`); the server records
 * that PR on the SESSION row and merges it through `codingSessions.mergePr`.
 * So a merge control now resolves to one of two targets, and every surface
 * that offers one keys its in-flight / failure state on [key].
 */
sealed interface MergeTarget {
    /** Stable identity for the merging + mergeErrors maps a surface keeps. */
    val key: String

    /** Merge through an issue (`issues.mergePr`) — single or batch PR. */
    data class Issue(val issueId: String) : MergeTarget {
        override val key: String get() = issueId
    }

    /** Merge the run's own PR (`codingSessions.mergePr`) — no issue is linked. */
    data class Session(val sessionId: String) : MergeTarget {
        override val key: String get() = "session:$sessionId"
    }
}

/** EXP-734: the run carries a pull request of its OWN that is still open. */
val CodingSessionEntity.hasOpenPr: Boolean
    get() = prState == DomainContract.prStateOpen && !prUrl.isNullOrEmpty()

/**
 * The merge control's target for one run, or null when there is nothing to
 * merge (no PR, or one already merged/closed).
 *
 * Order matters: an ISSUE-linked run always merges through its issue (the
 * server refuses `codingSessions.mergePr` on one — merging there completes the
 * issue), a batch run through the representative issue [batchPrIssue] the
 * caller resolved from the branch (EXP-535), and only an issueless run with no
 * batch resolution falls back to its own recorded PR.
 */
fun resolveMergeTarget(
    session: CodingSessionEntity,
    issue: IssueEntity?,
    batchPrIssue: IssueEntity? = null,
): MergeTarget? {
    if (session.issueId != null) {
        return issue
            ?.takeIf { it.prState == DomainContract.prStateOpen }
            ?.let { MergeTarget.Issue(it.id) }
    }
    batchPrIssue?.takeIf { it.prState == DomainContract.prStateOpen }?.let {
        return MergeTarget.Issue(it.id)
    }
    if (session.hasOpenPr) return MergeTarget.Session(session.id)
    return null
}

File: apps/ios/ExpCore/Sources/Domain/MergeTarget.swift
import Foundation

/// EXP-734: WHAT a run's Merge affordance merges through. An issue-linked run
/// (and a batch run, through its PR's representative issue) merges the ISSUE's
/// PR; an action or chat run's PR links no issue at all, so it merges through
/// the SESSION row the server stamped it on.
public enum MergeTarget: Equatable, Sendable {
    case issue(issueId: String)
    case session(sessionId: String)
}

/// The one rule every iOS merge surface applies (Agents rows, the steering
/// screen, Reviews) — mirrors web's `use-agents-data.ts` and the desktop.
public enum MergeTargetResolution {
    /// - Parameters:
    ///   - session: the run.
    ///   - issue: the run's own issue, when it has one (already observed).
    ///   - openBatchPrs: the team's open batch PRs, one representative issue
    ///     per distinct prUrl (`BatchPrResolution.openBatchPrs`).
    public static func resolve(
        session: CodingSessionEntity,
        issue: IssueEntity?,
        openBatchPrs: [IssueEntity]
    ) -> MergeTarget? {
        // An issue run merges its own issue's PR.
        if session.issueId != nil {
            if let issue, issue.prState == DomainContract.prStateOpen {
                return .issue(issueId: issue.id)
            }
            return nil
        }
        // A batch run carries no issue linkage — its PR resolves client-side
        // off the branch the pr_open flip stamped (EXP-535/545).
        if session.actionName == nil,
            session.status == DomainContract.codingSessionStatusInReview,
            let batchIssue = BatchPrResolution.resolve(
                sessionBranch: session.branch, openBatchPrs: openBatchPrs
            )
        {
            return .issue(issueId: batchIssue.id)
        }
        // Anything else merges through its OWN stamped PR — an action or chat
        // run's issue-less chore PR (EXP-734).
        if session.hasOpenPr {
            return .session(sessionId: session.id)
        }
        return nil
    }
}
